### PR TITLE
Refactoring getIsAuthorized()

### PR DESCRIPTION
**Pull request by profet23**
Refactoring `getIsAuthorized()` to wait until setup and refresh is complete before emitting.

### Previously:
`getIsAuthorized()` would emit its default value of `false` even before setup of the module had been completed. This lead to some very strange race conditions and numerous bug reports.

### With this change:
`getIsAuthorized()` will now:

1.  Wait for setup to complete
2.  Then:
     1. If silent renew _is not_ enabled:
          1. Return the result of authenticating any local tokens either successfully or unsuccessfully. 
     2. If silent renew _is_ enabled then SIMULTANEOUSLY:
          1. Authenticate locally stored tokens and return the result _ONLY_ if authentication is successful. _This action will never return if the locally stored tokens aren't successfully authenticated._
          2. Try and authenticate with the authority and return true or false based on outcome.
3. As soon as one of the above actions complete, the observable will emit and you'll have the authentication result.
4. This result is replayed to any late subscribers.

**While I think the impact will be minimal, this does create a fundamental change to an external facing API so care should be taken when upgrading.**

This issue has come up multiple times:
#257 
#291
#299

### CODE REVIEW

**Comment by damienbod** [comment]
@profet23 Thanks for this. After testing, I have problems with this change. I moved these changes to the Refactoring_getIsAuthorized branch for further testing.

I'm testing now to find out the problem.  When the silent renew is active, the oidc loops and produces incorrect warnings. I will test later to find out exactly why. 

If you want to see my test repo which fails:

https://github.com/damienbod/AspNet5IdentityServerAngularImplicitFlow/tree/dev

in the dev branch.

Greetings Damien

**Comment by damienbod** [comment]
@profet23 Looks like your changes are not the problem. Sorry with the latest version of the STS, the error: "consent_required" is now required per default, ans so the silent renew fails, because it is not possible to do a silent renew.

Will test and fix, but your changes look good to go.

Thanks and sorry for the last comment

**Comment by damienbod** [comment]
Have a different problem now :)
error: "login_required
The silent renew should not run until the user has logged in.

**Comment by damienbod** [comment]
Problem is that the this.refreshSession(); is getting called, even though the user has never logged in.

**Comment by damienbod** [comment]
@profet23 When I remove the  this.refreshSession(); from your code, everything works again. Can I remove this, or what was the reason you added it?

Greetings Damien

**Comment by profet23** [comment]
I get the error "login_required' as well. but that was intentional as the `getIsAuthorized()` simply returns false in that situation. And then I thought it would be up to the developer to kick off the authentication.

As an example I have this in a guard. The guard returns getIsAuthorized() and taps it. If the observable emits false then I navigate the user to the signin page of my app which kicks off the login process.

I'm using the silent renew to authenticate a user silently who already has logged into the auth server. For example I have my app running on multiple domains. If the user logs into the first domain, when they open another domain the app will retrieve credentials via silent renew. 

TLDR:  You'll get the error "login_required" but the observable should emit false in that situation and allow the developer to react.

**Comment by damienbod** [comment]
OK, I get a loop and the user cannot react. If I removed the this.refreshSession();, would it still work for you? This fixes my test env. 

Also the message 'OidcSecurityCheckSession pollServerSession session_state is blank' should be a debug log because this is a normal possibility. I will change it the logDebug, good?

And also thanks a million for you help.

Greetings Damien

**Comment by profet23** [comment]
@damienbod the `this.refreshSession()` is what kicks off emitting of the `onAuthorizationResult` event in the second part of the race.

It allows the app to silently authenticate a user who is already authenticated with the server. If they are not authenticated with the server, then the observable should (and I believe does) emit false and the developer needs to call authorize().

Without the `this.refreshSession()` you'll never be able to silently (via iframe) authenticate a user who is already authenticated. 

Maybe this isn't the correct workflow? Not sure. 

As for your loop, my guess is that something else in your app is subscribing either to the `onAuthorizationResult` event ?

**Comment by profet23** [comment]
But the refreshsession() will run only after 10+ seconds. We're looking to get the session refreshed right away as getIsAuthorized() will be used in an guards.

**Comment by damienbod** [comment]
@profet23 yes you are correct.  Reason I was looping, was because I was redirecting if not authorized, and the refresh function was called again. Navigated to the unauthorized, and not a reload, and then the solution worked, but I do 2 login requests with everything initial login, and 1 fails on the callback, the second is ok, because we use the correct nonce then. (1 request for the login, 1 for the refresh session call) Do you have these warnings, problem as well?

**Comment by profet23** [comment]
Yeah I think the refreshSession() is required because without it, getIsAuthorized() may never emit or at least may take an extended amount of time to emit.

While the error message isn't ideal, it is correct. A login is required, so the observable emits false and the developer can then react by calling authorize() if appropriate.

With this change (and adding back the refreshSession() call) a lot of logic should become easier for the developer in that they are mostly just listening for getIsAuthorized() emissions and reacting.

As for the checkForExistingTokensOnInitialOpen I have it doing that by checking if silent renew is configured. I figured if the user has silent renew configured they'd want to silently authenticate.

**Comment by damienbod** [comment]
@profet23 I will try to add it back. Just need to deactivate the silent renew refreshsession call, when the login authorize is running and then everything is fine.
